File: contrib/pg_tde/src/pg_tde_change_key_provider.c
#include "postgres_fe.h"

#include <stdarg.h>
#include <stdio.h>

#include "common/controldata_utils.h"
#include "common/logging.h"

#include "catalog/tde_global_space.h"
#include "catalog/tde_keyring.h"
#include "common/pg_tde_utils.h"
#include "pg_tde.h"

/* version string we expect back from pg_tde_change_key_provider */
#define PROGNAME "pg_tde_change_key_provider (PostgreSQL) " PG_VERSION "\n"

static void
help(void)
{
	puts("pg_tde_change_key_provider changes the configuration of a pg_tde key provider");
	puts("");
	puts("Usage:");
	puts("");
	puts("pg_tde_change_key_provider [-D <datadir>] <dbOid> <provider_name> <new_provider_type> <provider_parameters...>");
	puts("");
	puts("  Where <new_provider_type> can be file, vault-v2 or kmip");
	puts("");
	puts("Depending on the provider type, the complete parameter list is:");
	puts("");
	puts("pg_tde_change_key_provider [-D <datadir>] <dbOid> <provider_name> file <filename>");
	puts("pg_tde_change_key_provider [-D <datadir>] <dbOid> <provider_name> vault-v2 <url> <token_path> <mount_path> [<ca_path>]");
	puts("pg_tde_change_key_provider [-D <datadir>] <dbOid> <provider_name> kmip <host> <port> <cert_path> <key_path> [<ca_path>]");
	puts("");
	printf("Use dbOid %d for global key providers.\n", GLOBAL_DATA_TDE_OID);
	puts("");
	puts("WARNING:");
	puts("");
	puts("This tool only changes the values, without properly XLogging the changes, or validating that keys can be fetched using them. Only use it in case the database is inaccessible and can't be started.\n");
}

#define BUFFER_SIZE 1024

static bool
build_json(char *buffer, int count,...)
{
	va_list		args;
	char	   *ptr;

	va_start(args, count);

	ptr = strcat(buffer, "{");

	for (int i = 0; i < count; ++i)
	{
		/* TODO: no validation about the paramters at all... */
		/* not much we can do without a proper JSON library */
		/* If the JSON is incorrect, it will fail a bit later when */
		/* we try to backread it */
		const char *key = va_arg(args, const char *);
		const char *value = va_arg(args, const char *);

		bool		addQuotes = value == NULL || (value[0] != '{' && value[0] != '"');

		if (i != 0)
		{
			ptr = strcat(ptr, ",");
		}

		ptr = strcat(ptr, "\"");
		ptr = strcat(ptr, key);
		ptr = strcat(ptr, "\":");

		if (addQuotes)
		{
			ptr = strcat(ptr, "\"");
		}
		if (value != NULL)
		{
			ptr = strcat(ptr, value);
		}
		if (addQuotes)
		{
			ptr = strcat(ptr, "\"");
		}
		if (ptr - buffer > BUFFER_SIZE)
		{
			fprintf(stderr, "Error: Configuration too long.\n");
			return false;
		}
	}
	va_end(args);

	ptr = strcat(ptr, "}");

	if (ptr - buffer > BUFFER_SIZE)
	{
		fprintf(stderr, "Error: Configuration too long.\n");
		return false;
	}

	return true;
}

int
main(int argc, char *argv[])
{
	char	   *provider_name;
	char	   *new_provider_type;
	char	   *datadir = getenv("PGDATA");

	int			argstart = 0;

	char		json[BUFFER_SIZE * 2] = {0,};
	ControlFileData *controlfile;
	bool		crc_ok;
	char		tdedir[MAXPGPATH] = {0,};
	char	   *cptr = tdedir;
	bool		provider_found = false;
	KeyringProviderRecordInFile record;

	Oid			db_oid;

	pg_logging_init(argv[0]);
	pg_logging_set_level(PG_LOG_WARNING);
	set_pglocale_pgservice(argv[0], PG_TEXTDOMAIN("pg_tde_change_key_provider"));

	if (argc > 1)
	{
		if (strcmp(argv[1], "--help") == 0 || strcmp(argv[1], "-?") == 0)
		{
			help();
			exit(0);
		}
		if (strcmp(argv[1], "--version") == 0 || strcmp(argv[1], "-V") == 0)
		{
			puts("pg_tde_change_key_provider (PostgreSQL) " PG_VERSION);
			exit(0);
		}
	}

	if (argc > 3)
	{
		if (strcmp(argv[1], "-D") == 0)
		{
			datadir = argv[2];
		}
		argstart += 2;
	}

	if (datadir == NULL || strlen(datadir) == 0)
	{
		help();
		puts("\n");
		fprintf(stderr, "Error: Data directory missing.\n");
		exit(1);
	}

	if (argc - argstart <= 3)
	{
		help();
		exit(1);
	}

	db_oid = atoi(argv[1 + argstart]);
	provider_name = argv[2 + argstart];
	new_provider_type = argv[3 + argstart];

	if (strcmp("file", new_provider_type) == 0)
	{
		provider_found = true;

		if (argc - argstart != 5)
		{
			help();
			puts("\n");
			fprintf(stderr, "Error: wrong number of arguments.\n");
			exit(1);
		}

		if (!build_json(json, 1, "path", argv[4 + argstart]))
		{
			exit(1);
		}
	}

	if (strcmp("vault-v2", new_provider_type) == 0)
	{
		provider_found = true;

		if (argc - argstart != 7 && argc - argstart != 8)
		{
			help();
			puts("\n");
			fprintf(stderr, "Error: wrong number of arguments.\n");
			exit(1);
		}

		if (!build_json(json, 4,
						"url", argv[4 + argstart],
						"tokenPath", argv[5 + argstart],
						"mountPath", argv[6 + argstart],
						"caPath", (argc - argstart > 7 ? argv[7 + argstart] : "")))
		{
			exit(1);
		}
	}

	if (strcmp("kmip", new_provider_type) == 0)
	{
		provider_found = true;

		if (argc - argstart != 8 && argc - argstart != 9)
		{
			help();
			puts("\n");
			fprintf(stderr, "Error: wrong number of arguments.\n");
			exit(1);
		}

		if (!build_json(json, 5,
						"host", argv[4 + argstart],
						"port", argv[5 + argstart],
						"caPath", (argc - argstart > 8 ? argv[8 + argstart] : ""),
						"certPath", argv[6 + argstart],
						"keyPath", argv[7 + argstart]))
		{
			exit(1);
		}
	}

	if (!provider_found)
	{
		help();
		puts("\n");
		fprintf(stderr, "Error: Unknown provider type: %s\n.", new_provider_type);
		exit(1);
	}

	/*
	 * Check if cluster is running.  This way we can be sure we have no
	 * concurrent modifcations of the key providers.  Note that this doesn't
	 * guard against someone starting the cluster concurrently.
	 */
	controlfile = get_controlfile(datadir, &crc_ok);
	if (!crc_ok)
		pg_fatal("pg_control CRC value is incorrect");

	if (controlfile->state != DB_SHUTDOWNED &&
		controlfile->state != DB_SHUTDOWNED_IN_RECOVERY)
		pg_fatal("cluster must be shut down");

	cptr = strcat(cptr, datadir);
	cptr = strcat(cptr, "/");
	cptr = strcat(cptr, PG_TDE_DATA_DIR);
	pg_tde_set_data_dir(tdedir);

	if (get_keyring_info_file_record_by_name(provider_name, db_oid, &record) == false)
	{
		fprintf(stderr, "Error: provider not found\n.");
		exit(1);
	}

	record.provider.provider_type = get_keyring_provider_from_typename(new_provider_type);
	memset(record.provider.options, 0, sizeof(record.provider.options));
	strncpy(record.provider.options, json, sizeof(record.provider.options));

	write_key_provider_info(&record, false);

	printf("Key provider updated successfully!\n");

	return 0;
}

File: contrib/pg_tde/t/pg_tde_change_key_provider.pl
use strict;
use warnings FATAL => 'all';
use PostgreSQL::Test::Cluster;
use PostgreSQL::Test::Utils;
use Test::More;

use JSON;

my $node = PostgreSQL::Test::Cluster->new('main');
$node->init;
$node->append_conf('postgresql.conf', q{shared_preload_libraries = 'pg_tde'});
$node->start;

$node->safe_psql('postgres', q{CREATE EXTENSION pg_tde});
$node->safe_psql('postgres',
	q{SELECT pg_tde_add_global_key_provider_file('global-provider', '/tmp/pg_tde_change_key_provider-global')}
);
$node->safe_psql('postgres',
	q{SELECT pg_tde_add_database_key_provider_file('database-provider', '/tmp/pg_tde_change_key_provider-database')}
);
my $db_oid = $node->safe_psql('postgres',
	q{SELECT oid FROM pg_catalog.pg_database WHERE datname = 'postgres'});
my $options;

my $token_file = "${PostgreSQL::Test::Utils::tmp_check}/vault_token";
append_to_file($token_file, 'DUMMY');

$node->stop;

command_like(
	[
		'pg_tde_change_key_provider',
		'-D' => $node->data_dir,
		$db_oid,
		'database-provider',
		'file',
		'/tmp/pg_tde_change_key_provider-database-2',
	],
	qr/Key provider updated successfully!/,
	'updates key provider to file type');

$node->start;

is( $node->safe_psql(
		'postgres',
		q{SELECT type FROM pg_tde_list_all_database_key_providers() WHERE name = 'database-provider'}
	),
	'file',
	'provider type is set to file');

$options = decode_json(
	$node->safe_psql(
		'postgres',
		q{SELECT options FROM pg_tde_list_all_database_key_providers() WHERE name = 'database-provider'}
	));
is( $options->{path},
	'/tmp/pg_tde_change_key_provider-database-2',
	'path is set correctly for file provider');

$node->stop;

command_like(
	[
		'pg_tde_change_key_provider',
		'-D' => $node->data_dir,
		$db_oid,
		'database-provider',
		'vault-v2',
		'https://vault-server.example:8200/',
		'mount-path',
		$token_file,
		'/tmp/ca_path',
	],
	qr/Key provider updated successfully!/,
	'updates key provider to vault-v2 type with https');

$node->start;

is( $node->safe_psql(
		'postgres',
		q{SELECT type FROM pg_tde_list_all_database_key_providers() WHERE name = 'database-provider'}
	),
	'vault-v2',
	'provider type is set to vault-v2');

$options = decode_json(
	$node->safe_psql(
		'postgres',
		q{SELECT options FROM pg_tde_list_all_database_key_providers() WHERE name = 'database-provider'}
	));
is( $options->{url},
	'https://vault-server.example:8200/',
	'url is set correctly for vault-v2 provider');
is($options->{mountPath}, 'mount-path',
	'mount path is set correctly for vault-v2 provider');
is($options->{tokenPath}, $token_file,
	'tokenPath is set correctly for vault-v2 provider');
is($options->{caPath}, '/tmp/ca_path',
	'CA path is set correctly for vault-v2 provider');

$node->stop;

command_like(
	[
		'pg_tde_change_key_provider',
		'-D' => $node->data_dir,
		$db_oid,
		'database-provider',
		'vault-v2',
		'http://vault-server.example:8200/',
		'mount-path-2',
		$token_file,
	],
	qr/Key provider updated successfully!/,
	'updates key provider to vault-v2 type with http');

$node->start;

is( $node->safe_psql(
		'postgres',
		q{SELECT type FROM pg_tde_list_all_database_key_providers() WHERE name = 'database-provider'}
	),
	'vault-v2',
	'provider type is set to vault-v2');

$options = decode_json(
	$node->safe_psql(
		'postgres',
		q{SELECT options FROM pg_tde_list_all_database_key_providers() WHERE name = 'database-provider'}
	));
is( $options->{url},
	'http://vault-server.example:8200/',
	'url is set correctly for vault-v2 provider');
is($options->{mountPath}, 'mount-path-2',
	'mount path is set correctly for vault-v2 provider');
is($options->{tokenPath}, $token_file,
	'tokenPath is set correctly for vault-v2 provider');
is($options->{caPath}, '', 'CA path is set correctly for vault-v2 provider');

$node->stop;

command_like(
	[
		'pg_tde_change_key_provider',
		'-D' => $node->data_dir,
		$db_oid,
		'database-provider',
		'kmip',
		'kmip-server.example',
		'12345',
		'/tmp/cert_path',
		'/tmp/key_path',
		'/tmp/ca_path',
	],
	qr/Key provider updated successfully!/,
	'updates key provider to kmip type');

$node->start;

is( $node->safe_psql(
		'postgres',
		q{SELECT type FROM pg_tde_list_all_database_key_providers() WHERE name = 'database-provider'}
	),
	'kmip',
	'provider type is set to kmip');

$options = decode_json(
	$node->safe_psql(
		'postgres',
		q{SELECT options FROM pg_tde_list_all_database_key_providers() WHERE name = 'database-provider'}
	));
is($options->{host}, 'kmip-server.example',
	'host is set correctly for kmip provider');
is($options->{port}, '12345', 'port is set correctly for kmip provider');
is($options->{certPath}, '/tmp/cert_path',
	'client cert path is set correctly for kmip provider');
is($options->{keyPath}, '/tmp/key_path',
	'client cert key path is set correctly for kmip provider');
is($options->{caPath}, '/tmp/ca_path',
	'CA path is set correctly for kmip provider');

$node->stop;

command_like(
	[
		'pg_tde_change_key_provider',
		'-D' => $node->data_dir,
		'1664',
		'global-provider',
		'vault-v2',
		'http://vault-server.example:8200/',
		'mount-path',
		$token_file,
		'/tmp/ca_path',
	],
	qr/Key provider updated successfully!/,
	'updates key provider to vault-v2 type for global provider');

$node->start;

is( $node->safe_psql(
		'postgres',
		q{SELECT type FROM pg_tde_list_all_global_key_providers() WHERE name = 'global-provider'}
	),
	'vault-v2',
	'provider type is set to vault-v2 for global provider');

$options = decode_json(
	$node->safe_psql(
		'postgres',
		q{SELECT options FROM pg_tde_list_all_global_key_providers() WHERE name = 'global-provider'}
	));
is( $options->{url},
	'http://vault-server.example:8200/',
	'options are updated for global provider');

$node->stop;

command_fails_like(
	[
		'pg_tde_change_key_provider',
		'-D' => '/non/existing/path',
		'1664',
		'global-provider',
		'file',
		'/tmp/file',
	],
	qr{pg_tde_change_key_provider: error: could not open file "/non/existing/path/global/pg_control" for reading: No such file or directory},
	'gives error on incorrect data dir');

$node->start;
command_fails_like(
	[
		'pg_tde_change_key_provider',
		'-D' => $node->data_dir,
		'1664',
		'global-provider',
		'file',
		'/tmp/file',
	],
	qr/pg_tde_change_key_provider: error: cluster must be shut down/,
	'gives error on if cluster is running');
$node->stop;

command_fails_like(
	[
		'pg_tde_change_key_provider',
		'-D' => $node->data_dir,
		'12345678',
		'global-provider',
		'file',
		'/tmp/file',
	],
	qr/error: could not open tde file "[^"]+": No such file or directory/,
	'gives error on unknown database oid');

command_fails_like(
	[
		'pg_tde_change_key_provider',
		'-D' => $node->data_dir,
		'1664',
		'incorrect-global-provider',
		'file',
		'/tmp/file',
	],
	qr/error: provder "incorrect-global-provider" not found for database 1664/,
	'gives error on unknown key provider');

command_fails_like(
	[
		'pg_tde_change_key_provider',
		'-D' => $node->data_dir,
		'1664',
		'global-provider',
		'incorrect-provider-type',
	],
	qr/error: unknown provider type "incorrect-provider-type"/,
	'gives error on unknown provider type');

command_fails_like(
	[
		'pg_tde_change_key_provider',
		'-D' => $node->data_dir,
		'1664',
		'global-provider',
		'file',
	],
	qr/error: wrong number of arguments for "file"/,
	'gives error on missing arguments for file provider');

command_fails_like(
	[
		'pg_tde_change_key_provider',
		'-D' => $node->data_dir,
		'1664',
		'global-provider',
		'kmip',
	],
	qr/error: wrong number of arguments for "kmip"/,
	'gives error on missing arguments for kmip provider');

command_fails_like(
	[
		'pg_tde_change_key_provider',
		'-D' => $node->data_dir,
		'1664',
		'global-provider',
		'vault-v2',
	],
	qr/error: wrong number of arguments for "vault-v2"/,
	'gives error on missing arguments for vault-v2 provider');

done_testing();
